File: replication package/scripts/1_main_paper/figures/fig4.py
# Emotion and Reason in Political LanguageL: Replication Package
# Gennaro and Ash

# Description:
# - Replicates Figure 4: Emotionality by Topic Over Time.

###################################
#     Modules                   ###
###################################

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import random

random.seed(10)

###################################
#     Working Directory         ###
###################################

# indicate here the path to the replication package
wd = "/Users/glgennaro/Dropbox (Personal)/Progetti/Ash_Gennaro/1 EJ paper/EJ-RR/20210282_final_submission"

wd_data = wd + '/3 Replication Package/data/1_main_datasets'  # set the data directory
wd_results = wd + '/3 Replication Package/results/main_paper'  # set the results directory
wd_aux = wd + '/3 Replication Package/data/3_auxiliary_data'

###################################
#     Upload dataset            ###
###################################

df = pd.read_csv(wd_data + '/main_dataset.csv')

# Add demeaned score and adjust come vars
df = df.rename(columns={"topic1_new": "topic_num"})
df['topic_num'] = pd.to_numeric(df['topic_num'], downcast='integer')
df['score_demeaned'] = df['score'] - df['score'].groupby(df['congress']).transform('mean')


###################################
#     Upload topic labels       ###
###################################

topic_labels = pd.read_csv(wd_aux + '/topics_numbers.csv', header=None)
topic_labels = list(topic_labels[0])
topic_labels = [[topic_labels[x], topic_labels[x + 1], topic_labels[x + 2]]
                for x in range(0, len(topic_labels) - 1, 3)]
topic_labels = pd.DataFrame(topic_labels, columns=['topic_num', 'topic', 'theme'])
topic_labels['topic_num'] = pd.to_numeric(topic_labels['topic_num'], downcast='integer')
topic_labels[['topic_broad', 'topic_detail']] = topic_labels['topic'].str.split(' - ', 1, expand=True)


topic_labels['topic_macro'] = ''
topic_labels['topic_macro'].iloc[topic_labels['topic_broad'].isin(['Economic Policy', 'Fiscal Policy', 'Monetary Policy'])] = 'Economy'
topic_labels['topic_macro'].iloc[topic_labels['topic_broad'].isin(['Governance'])] = 'Governance'
topic_labels['topic_macro'].iloc[topic_labels['topic_broad'].isin(['Immigration', 'Social Issues'])] = 'Society'
topic_labels['topic_macro'].iloc[topic_labels['topic_broad'].isin(['Foreign Policy'])] = 'Foreign Affairs'
topic_labels['topic_macro'].iloc[topic_labels['topic_broad'].isin(['Procedure'])] = 'Procedure'
topic_labels['topic_macro'].iloc[topic_labels['topic_broad'].isin(['National Narrative'])] = 'National Narrative'
topic_labels['topic_macro'].iloc[topic_labels['topic_broad'].isin(['Party Politics', 'Tribute'])] = 'Party Politics'


df = pd.merge(df, topic_labels, on='topic_num', how='left')

df = df[df.congress >= 56]  # Starting on 1900



###################
#     Panel A   ###
###################

data = df[['congress', 'speech_year', 'topic_macro', 'score']]
lab = data[['congress', 'speech_year']].groupby('congress').first().rename(columns={'speech_year': 'labyear'})

data = pd.merge(data, lab, on='congress', how='left')

t1 = data.groupby(['topic_macro', 'labyear'])['score'].mean().reset_index()
t1.columns = ['topic', 'labyear', 'score']
t2 = data.groupby(['topic_macro', 'labyear'])['score'].sem().reset_index()
t2.columns = ['topic', 'labyear', 'score_std']

topics = pd.merge(t1, t2, on=['topic', 'labyear'], how='left')
topics = topics[topics.topic != '']

# Plot
labels = list(set(topics['topic']))
fig, ax = plt.subplots(figsize=(30, 15))
for label in labels:
    x = topics['labyear'][topics.topic == label]  # immigration
    y = topics['score'][topics.topic == label]
    if label=='Economy':
        line1, = ax.plot(x, y, label=label, linewidth=4)
    else:
	    line1, = ax.plot(x, y, label=label)
plt.yticks(fontsize=20)
plt.xticks(np.arange(1900, 2015, 4), fontsize=20, rotation=90)
ax.legend(loc='upper left', prop={'size': 24})
plt.title('Panel A', fontsize=40, loc="left")
plt.grid()
plt.savefig(wd_results + '/fig4a.png')




###################
#     Panel B   ###
###################



data = df[['congress', 'speech_year', 'topic_broad', 'score']]
lab = data[['congress', 'speech_year']].groupby('congress').first().rename(columns={'speech_year': 'labyear'})
data = pd.merge(data, lab, on='congress', how='left')


t1 = data.groupby(['topic_broad', 'labyear'])['score'].mean().reset_index()
t1.columns = ['topic', 'labyear', 'score']
t2 = data.groupby(['topic_broad', 'labyear'])['score'].sem().reset_index()
t2.columns = ['topic', 'labyear', 'score_std']

topics = pd.merge(t1, t2, on=['topic', 'labyear'], how='left')

# Plot
labels = ['Economic Policy', 'Monetary Policy', 'Fiscal Policy']
fig, ax = plt.subplots(figsize=(30, 15))
for l in labels:
	x = topics['labyear'][topics.topic==l]  # immigration
	y = topics['score'][topics.topic==l]
	#s = topics['score_std'][topics.topic==l]
	#ax.fill_between(x, (y-s), (y+s), alpha=.2)
	line1, = ax.plot(x, y, label=l)
plt.yticks(fontsize=20)
plt.title('Panel B', fontsize=40, loc="left")
plt.xticks(np.arange(1900, 2015, 4), fontsize=20, rotation=90)
plt.grid()

ax.legend(loc='upper left', prop={'size': 24})
plt.savefig(wd_results + '/fig4b.png')

File: R_version/fig4.R
# Install required libraries if not already installed
install.packages(c("ggplot2", "dplyr", "readr", "tidyr"))

# Load libraries
library(ggplot2)
library(dplyr)
library(readr)
library(tidyr)
# Set working directory
wd <- "/path/to/your/directory"  # Replace with your actual directory

# Set data and results paths
wd_data <- paste0(wd, "/3 Replication Package/data/1_main_datasets")
wd_results <- paste0(wd, "/3 Replication Package/results/main_paper")
wd_aux <- paste0(wd, "/3 Replication Package/data/3_auxiliary_data")

# Load dataset
df <- read_csv(paste0(wd_data, "/main_dataset.csv"))
# Add demeaned score and adjust some variables
df <- df %>%
  rename(topic_num = topic1_new) %>%
  mutate(
    topic_num = as.integer(topic_num),
    score_demeaned = score - ave(score, congress, FUN = mean)
  )

# Load and process topic labels
topic_labels <- read_csv(paste0(wd_aux, "/topics_numbers.csv"), col_names = FALSE)
topic_labels <- topic_labels %>%
  mutate(
    topic_num = as.integer(X1),
    topic = X2,
    theme = X3
  ) %>%
  separate(topic, into = c("topic_broad", "topic_detail"), sep = " - ", fill = "right") %>%
  mutate(
    topic_macro = case_when(
      topic_broad %in% c("Economic Policy", "Fiscal Policy", "Monetary Policy") ~ "Economy",
      topic_broad == "Governance" ~ "Governance",
      topic_broad %in% c("Immigration", "Social Issues") ~ "Society",
      topic_broad == "Foreign Policy" ~ "Foreign Affairs",
      topic_broad == "Procedure" ~ "Procedure",
      topic_broad == "National Narrative" ~ "National Narrative",
      topic_broad %in% c("Party Politics", "Tribute") ~ "Party Politics",
      TRUE ~ NA_character_
    )
  )

# Merge topic labels with main dataset
df <- left_join(df, topic_labels, by = "topic_num")

# Filter dataset to start from 1900 (Congress 56)
df <- df %>%
  filter(congress >= 56)
# Panel A - Plotting
data <- df %>%
  select(congress, speech_year, topic_macro, score)

# Extract the first speech year for each congress
lab <- data %>%
  group_by(congress) %>%
  summarize(labyear = first(speech_year))

# Merge to get labyear column in the dataset
data <- left_join(data, lab, by = "congress")

# Compute mean score by topic and year
t1 <- data %>%
  group_by(topic = topic_macro, labyear) %>%
  summarize(score = mean(score, na.rm = TRUE))

# Compute standard error of the mean
t2 <- data %>%
  group_by(topic = topic_macro, labyear) %>%
  summarize(score_std = sd(score, na.rm = TRUE) / sqrt(n()))

# Merge the data
topics <- left_join(t1, t2, by = c("topic", "labyear"))
topics <- topics %>%
  filter(topic != "")

# Plot
labels <- unique(topics$topic)
ggplot() +
  geom_line(data = topics, aes(x = labyear, y = score, color = topic, size = (topic == "Economy")), show.legend = TRUE) +
  scale_size_manual(values = c(1, 2)) +
  scale_color_manual(values = c("red", "blue", "green", "purple", "orange", "pink", "brown", "yellow")) +
  scale_x_continuous(breaks = seq(1900, 2015, 4)) +
  theme_minimal(base_size = 20) +
  theme(
    axis.text.x = element_text(angle = 90, size = 20),
    axis.text.y = element_text(size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_text(size = 24),
    plot.title = element_text(size = 40, hjust = 0),
    plot.subtitle = element_text(size = 25, hjust = 0)
  ) +
  labs(
    title = "Panel A",
    x = "Year",
    y = "Average Emotionality",
    subtitle = "Emotionality by Topic over Time"
  ) +
  theme(legend.position = "upper left") +
  ggsave(paste0(wd_results, "/fig4a.png"), width = 30, height = 15)
# Panel B - Plotting
data <- df %>%
  select(congress, speech_year, topic_broad, score)

# Extract the first speech year for each congress
lab <- data %>%
  group_by(congress) %>%
  summarize(labyear = first(speech_year))

# Merge to get labyear column in the dataset
data <- left_join(data, lab, by = "congress")

# Compute mean score by topic and year
t1 <- data %>%
  group_by(topic = topic_broad, labyear) %>%
  summarize(score = mean(score, na.rm = TRUE))

# Compute standard error of the mean
t2 <- data %>%
  group_by(topic = topic_broad, labyear) %>%
  summarize(score_std = sd(score, na.rm = TRUE) / sqrt(n()))

# Merge the data
topics <- left_join(t1, t2, by = c("topic", "labyear"))

# Select relevant topics for Panel B
labels <- c('Economic Policy', 'Monetary Policy', 'Fiscal Policy')

# Plot
ggplot() +
  geom_line(data = topics %>% filter(topic %in% labels), aes(x = labyear, y = score, color = topic)) +
  scale_x_continuous(breaks = seq(1900, 2015, 4)) +
  scale_color_manual(values = c("red", "green", "blue")) +
  theme_minimal(base_size = 20) +
  theme(
    axis.text.x = element_text(angle = 90, size = 20),
    axis.text.y = element_text(size = 20),
    legend.text = element_text(size = 20),
    legend.title = element_text(size = 24),
    plot.title = element_text(size = 40, hjust = 0),
    plot.subtitle = element_text(size = 25, hjust = 0)
  ) +
  labs(
    title = "Panel B",
    x = "Year",
    y = "Average Emotionality",
    subtitle = "Emotionality by Subtopic over Time"
  ) +
  theme(legend.position = "upper left") +
  ggsave(paste0(wd_results, "/fig4b.png"), width = 30, height = 15)
